testing/example_scripts/dataclasses/test_compare_recursive_dataclasses.py

```diff
@@ -1,5 +1,6 @@
 from dataclasses import dataclass
```

Comment:
I think the names here are bit too long/unwieldy for the tests, and also it doesn't test the situation perfectly.

I played with it a bit and I think this file would be a good replacement:

```py
from dataclasses import dataclass


@dataclass
class S:
    a: int
    b: str


@dataclass
class C:
    c: S
    d: S


@dataclass
class C2:
    e: C
    f: S


@dataclass
class C3:
    g: S
    h: C2
    i: str
    j: str


def test_recursive_dataclasses():
    left = C3(
        S(10, "ten"), C2(C(S(1, "one"), S(2, "two")), S(2, "three")), "equal", "left",
    )
    right = C3(
        S(20, "xxx"), C2(C(S(1, "one"), S(2, "yyy")), S(3, "three")), "equal", "right",
    )

    assert left == right
```

And IMO this would be the desirable output (but let me know if you think differently):

```
E       AssertionError: assert C3(g=S(a=10, b='ten'), h=C2(e=C(c=S(a=1, b='one'), d=S(a=2, b='two')), f=S(a=2, b='three')), i='equal', j='left') == C3(g=S(a=20, b='xxx'), h=C2(e=C(c=S(a=1, b='one'), d=S(a=2, b='yyy')), f=S(a=3, b='three')), i='equal', j='right')
E         
E         Matching attributes:
E         ['i']
E         Differing attributes:
E         ['g', 'h', 'j']
E         
E         Drill down into differing attribute g:
E           g: S(a=10, b='ten') != S(a=20, b='xxx')
E           
E           Differing attributes:
E           ['a', 'b']
E           
E           Drill down into differing attribute a:
E             a: 10 != 20
E             +10
E             -20
E           
E           Drill down into differing attribute b:
E             b: 'ten' != 'xxx'
E             - xxx
E             + ten
E         
E         Drill down into differing attribute h:
E           h: C2(e=C(c=S(a=1, b='one'), d=S(a=2, b='two')), f=S(a=2, b='three')) != C2(e=C(c=S(a=1, b='one'), d=S(a=2, b='yyy')), f=S(a=3, b='three'))
E           
E           Differing attributes:
E           ['e', 'f']
E           
E           Drill down into differing attribute e:
E             e: C(c=S(a=1, b='one'), d=S(a=2, b='two')) != C(c=S(a=1, b='one'), d=S(a=2, b='yyy'))
E             
E             Matching attributes:
E             ['c']
E             Differing attributes:
E             ['d']
E             
E             Drill down into differing attribute d:
E               d: S(a=2, b='two') != S(a=2, b='yyy')
E               
E               Matching attributes:
E               ['a']
E               Differing attributes:
E               ['b']
E               
E               Drill down into differing attribute b:
E                 b: 'two' != 'yyy'
E                 - yyy
E                 + two
E           
E           Drill down into differing attribute f:
E             f: S(a=2, b='three') != S(a=3, b='three')
E             
E             Matching attributes:
E             ['b']
E             Differing attributes:
E             ['a']
E             
E             Drill down into differing attribute a:
E               a: 2 != 3
E               +2
E               -3
E         
E         Drill down into differing attribute j:
E           j: 'left' != 'right'
E           - right
E           + left
```
```

Reply:
@bluetech does this output look OK to you? Difference to your proposed output is that:
- the lines where differing attributes are being compared (like: `g: S(a=10, b='ten') != S(a=20, b='xxx')`) 
- and the lines where differing fields of "simple" data structures are being outputted (like `+10`)

are on the same level as the expression `Drill down into differing attribute <attr_name>`. I could put the first type of lines on an 1x more nested level, but with the second type it would be more complicated, I guess: we either need to check the output of `_compare_eq_any()`, or pass the `indentation_level` again to all comparison functions. Or am I wrong and miss something?

Current output (not pushed yet):

```
E       AssertionError: assert C3(g=S(a=10, b='ten'), h=C2(e=C(c=S(a=1, b='one'), d=S(a=2, b='two')), f=S(a=2, b='three')), i='equal', j='left') == C3(g=S(a=20, b='xxx'), h=C2(e=C(c=S(a=1, b='one'), d=S(a=2, b=
'yyy')), f=S(a=3, b='three')), i='equal', j='right')
E         
E           Matching attributes:
E           ['i']
E           Differing attributes:
E           ['g', 'h', 'j']
E         
E           Drill down into differing attribute g:
E           g: S(a=10, b='ten') != S(a=20, b='xxx')
E           
E             Differing attributes:
E             ['a', 'b']
E           
E             Drill down into differing attribute a:
E             a: 10 != 20
E             +10
E             -20
E           
E             Drill down into differing attribute b:
E             b: 'ten' != 'xxx'
E             - xxx
E             + ten
E         
E           Drill down into differing attribute h:
E           h: C2(e=C(c=S(a=1, b='one'), d=S(a=2, b='two')), f=S(a=2, b='three')) != C2(e=C(c=S(a=1, b='one'), d=S(a=2, b='yyy')), f=S(a=3, b='three'))
E           
E             Differing attributes:
E             ['e', 'f']
E           
E             Drill down into differing attribute e:
E             e: C(c=S(a=1, b='one'), d=S(a=2, b='two')) != C(c=S(a=1, b='one'), d=S(a=2, b='yyy'))
E             
E               Matching attributes:
E               ['c']
E               Differing attributes:
E               ['d']
E             
E               Drill down into differing attribute d:
E               d: S(a=2, b='two') != S(a=2, b='yyy')
E               
E                 Matching attributes:
E                 ['a']
E                 Differing attributes:
E                 ['b']
E               
E                 Drill down into differing attribute b:
E                 b: 'two' != 'yyy'
E                 - yyy
E                 + two
E           
E             Drill down into differing attribute f:
E             f: S(a=2, b='three') != S(a=3, b='three')
E             
E               Matching attributes:
E               ['b']
E               Differing attributes:
E               ['a']
E             
E               Drill down into differing attribute a:
E               a: 2 != 3
E               +2
E               -3
E         
E           Drill down into differing attribute j:
E           j: 'left' != 'right'
E           - right
E           + left
```

@nicoddemus  for your info.

Reply:
That's already pretty good!

I do prefer my version because for visual consistency I think it's better if each "Drill down" actually drills down instead of sometimes yes and sometimes no. The idea is that the user gets the idea just from the visual tree structure even before reading the details.

> I could put the first type of lines on an 1x more nested level

Right, this one line you can just indent.

> but with the second type it would be more complicated, I guess: we either need to check the output of _compare_eq_any(), or pass the indentation_level again to all comparison functions. Or am I wrong and miss something?

If we indent all of the lines of the recursive `_compare_eq_any` (in `_compare_eq_cls`, not by passing indent level) then it should work out well unless I'm missing something. Is that what you mean by "check the output of `_compare_eq_any()`"?

Reply:
> I do prefer my version because for visual consistency I think it's better if each "Drill down" actually drills down instead of sometimes yes and sometimes no. The idea is that the user gets the idea just from the visual tree structure even before reading the details.

I agree, because then the `Drill down into` acts like a header.

Reply:
@bluetech Not exactly what I mean. 

I already indent all the lines of the recursive `_compare_eq_any` (in `_compare_eq_cls`). But since in our example the last values in the hierarchy are differing integer / string values (not dataclasses / attrs), we indent them with same level, as the `Drill down into`. If we indent everything from `_compare_eq_any` with double level, then it will lead to a new problem: analysis of the difference (e.g. `Differing attributes:`) will be on diff. level of nestedness, than e.g. `g: S(a=10, b='ten') != S(a=20, b='xxx')`. 

So, basically, we can't have same indentation level for both types of lines, nor different indentation level. That's why I said, that I see only one possibility to fix this: check the output of `_compare_eq_any()`. By that I mean the following: if the line starts with `"+"` or `"-"`, double the indentation for these lines.

Reply:
I will give an example. This is what happens when I indent both types of lines with double level:

```
E       AssertionError: assert C3(g=S(a=10, b='ten'), h=C2(e=C(c=S(a=1, b='one'), d=S(a=2, b='two')), f=S(a=2, b='three')), i='equal', j='left') == C3(g=S(a=20, b='xxx'), h=C2(e=C(c=S(a=1, b='one'), d=S(a=2, b=
'yyy')), f=S(a=3, b='three')), i='equal', j='right')
E         
E           Matching attributes:
E           ['i']
E           Differing attributes:
E           ['g', 'h', 'j']
E         
E           Drill down into differing attribute g:
E             g: S(a=10, b='ten') != S(a=20, b='xxx')
E             
E               Differing attributes:
E               ['a', 'b']
```

Reply:
In general I don't see a reason for this to happen. Can you push your latest version, then we can figure it out? (The tests don't need to pass in the meanwhile).

Reply:
@bluetech did it

Reply:
Thanks! I left comments on the changes I think we should make.